src/python/pants/backend/python/util_rules/pex.py

```diff
@@ -587,37 +594,85 @@ def _validate_metadata(
     if validation:
         return
 
-    message_parts = [
-        f"Invalid lockfile for PEX request `{request.output_filename}`.",
-        "\n",
-    ]
+    def message_parts() -> Iterable[str]:
+
+        tool_name = request.options_scope_name
+        uses_source_plugins = tool_name in ["mypy", "pylint"]
+        uses_project_interpreter_constraints = tool_name in [
+            "bandit",
+            "flake8",
+            "pylint",
+            "setuptools",
+            "ipython",
+            "pytest",
+        ]
```

Comment:
To be scalable and work with plugins, this needs to come from call sites imo. Maybe subclass `LockfileContent`  and `Lockfile`?

```python
class ToolDefaultLockfile(LockfileContent):
    option_scope: str
    uses_source_plugins: bool
    uses_project_interpreter_constraints: bool

class ToolCustomLockfile(Lockfile):
    option_scope: str
    uses_source_plugins: bool
    uses_project_interpreter_constraints: bool
```

`PythonToolBase.pex_requirements()` will set those up properly. Likely the callers of `.pex_requirements()` would need to override the default to indicate that tool's behavior. For example, you could rename this `extra_requirements` to be `requirements_from_source_plugins`, so if it's set you know to set `uses_source_plugins=True`:

https://github.com/pantsbuild/pants/blob/26caaf8a7f2890f5ed25bcd5c0c931476654d5c3/src/python/pants/backend/python/subsystems/python_tool_base.py#L126-L130

This allows you to remove `options_scope_name` from the generic `PexRequest`, so it's instead set in `ToolDefaultLockfile` and `ToolCustomLockfile`.

In this method, if you encounter a generic `LockfileContent` or `Lockfile`, you can punt on those error messages being good. We'll refactor things as part of multiple user lockfiles.

Reply:
That all makes sense

Reply:
It can be changed safely. Generally, the Plugin API is not stable (but we try to minimize churn), and it's only been in today's rc0. I doubt anyone outside of Pants is consuming it yet. There should be only two callers internally, both using it for source plugins.

--

For interpreter constraints, I was pondering the heuristic of `register_interpreter_constraints = False` meaning that you should set `uses_interpreter_constraints=True`. I think that's actually safe to do, at least de facto. We don't have any tools that suport lockfiles yet have no notion of interpreter constraints, and that wouldn't work.

Reply:
On reflection (and correct me if I'm wrong): it doesn't matter if `requirements_from_source_plugins` is set, it matters that the option exists and it _could be set_, right?

Reply:
Ah true. The only challenge is that there is no standardized way to check if the option --source-plugins is set, and it only happens to be that MyPy and Pylint call the option the same thing. Maybe you could use a default of None, whereas those two tools will pass [] if there are no plugins, and you can use that to determine the difference? I think --source-plugins already defaults to [] because it's a list option.

Reply:
My current tack is to add a new `uses_requirements_from_source_plugins` property to `PythonToolBase`, and we can discuss the heuristic later